Prometheus backend

aleksandrzhiliaev opened the pull request:
This PR adds new Prometheus backend support:
```go

        operationWithoutLabels := bucket.NewMetricOperation("files", "file", "download")
	operation := bucket.NewMetricOperation("files", "file", "upload")

	clientPrometheus, _ := stats.NewClient("prometheus://azh_test")
	defer clientPrometheus.Close()

	// track metric
	clientPrometheus.TrackMetric("a1", operationWithoutLabels)
	clientPrometheus.TrackMetric("a1", operationWithoutLabels)
	clientPrometheus.TrackMetric("a1", operationWithoutLabels)

	clientPrometheus.TrackMetric("a2", operation.WithLabels(map[string]string{"type": "png", "method": "post"}))
	clientPrometheus.TrackMetric("a2", operation.WithLabels(map[string]string{"type": "png", "method": "post"}))
	clientPrometheus.TrackMetric("a2", operation.WithLabels(map[string]string{"type": "jpg"}))
	clientPrometheus.TrackMetric("a2", operation.WithLabels(map[string]string{"method": "post"}))

	clientPrometheus.TrackMetricN("a3", operation.WithLabels(map[string]string{"type": "jpg"}), 123)
	clientPrometheus.TrackMetricN("a3", operation.WithLabels(map[string]string{"type": "jpg"}), 123)
	clientPrometheus.TrackMetricN("a3", operation.WithLabels(map[string]string{"method": "post", "type": "gif"}), 123)

	clientPrometheus.TrackMetricN("a4", operationWithoutLabels, 999)

	// track state
	clientPrometheus.TrackState("a5", operationWithoutLabels, 12345)
	clientPrometheus.TrackState("a6", operation.WithLabels(map[string]string{"method": "delete"}), 888)
	clientPrometheus.TrackState("a6", operation.WithLabels(map[string]string{"method": "delete"}), 2)

	// track operation
	clientPrometheus.TrackOperation("a7", operationWithoutLabels, nil, true)
	clientPrometheus.TrackOperation("a7", operationWithoutLabels, nil, true)
	clientPrometheus.TrackOperation("a7", operationWithoutLabels, nil, false)
	clientPrometheus.TrackOperation("a7", operationWithoutLabels, nil, false)

	clientPrometheus.TrackOperation("a8", operation.WithLabels(map[string]string{"method": "request"}), nil, true)
	clientPrometheus.TrackOperation("a8", operation.WithLabels(map[string]string{"type": "png"}), nil, false)

	// track operation with timer
	timing := clientPrometheus.BuildTimer().Start()
	r := rand.Intn(1000)
	time.Sleep(time.Duration(r) * time.Microsecond)
	clientPrometheus.TrackOperation("a9", operation.WithLabels(map[string]string{"type": "png"}), timing, false)

	clientPrometheus.TrackOperationN("a91", operationWithoutLabels, nil, 3, true)
	clientPrometheus.TrackOperationN("a92", operation.WithLabels(map[string]string{"method": "get", "type": "png"}), nil, 3, true)

	// track requests
	router := gin.Default()
	router.Use(NewStatsRequest(clientPrometheus))

	router.GET("/", func(c *gin.Context) {
		c.JSON(http.StatusOK, "I'm producing stats!")
	})
	router.GET("/test", func(c *gin.Context) {
		c.JSON(http.StatusOK, "I'm producing test stats!")
	})
	router.GET("/metrics", gin.WrapH(clientPrometheus.Handler()))

	router.Run(":8080")
```

TODO:
- [X] Timers support
- [x] Think about better interface for version 2.0

vgarvardt commented on client.go:
```diff
@@ -4,14 +4,15 @@ import (
-	"strings"
-
 	"github.com/hellofresh/stats-go/client"
+	"strings"
```
this lib if from stdlib, it should be groupped with other stdlib imports

vgarvardt commented on client/client.go:
```diff
@@ -19,24 +19,37 @@ type Client interface {
 	TrackMetric(section string, operation bucket.MetricOperation) Client
+
+	// TrackMetricWithLabels tracks custom metric with added labels, w/out ok/fail additional sections
+	TrackMetricWithLabels(section string, operation bucket.MetricOperation, labels map[string]string) Client
```
too many implementation-specific methods in the interface, maybe we need to rethink this a bit, maybe even by breaking interface, e.g. replace `bucket.MetricOperation` with the builder like we do for timer, need to think more about it

vgarvardt commented on client/log.go:
```diff
@@ -84,6 +84,17 @@ func (c *Log) TrackMetric(section string, operation bucket.MetricOperation) Clie
 
+// TrackMetricWithLabels tracks custom metric with added labels, w/out ok/fail additional sections
+func (c *Log) TrackMetricWithLabels(section string, operation bucket.MetricOperation, labels map[string]string) Client {
+	b := bucket.NewPlain(section, operation, true, c.unicode)
```
you could simply replace copy-pasted method body with `return c.TrackMetric(...)`

vgarvardt commented on client/log.go:
```diff
@@ -135,3 +157,8 @@ func (c *Log) SetHTTPRequestSection(section string) Client {
+
+// Handler returns metrics endpoint for prometheus backend
+func (c *Log) Handler() http.Handler {
+	return nil
```
we'd better return handler always - in this case handler that, for example, returns `405` status code

vgarvardt replied:
still valid =)

vgarvardt commented on client/prometheus.go:
```diff
@@ -0,0 +1,233 @@
+	"net/http"
+	"github.com/hellofresh/stats-go/bucket"
+	"github.com/hellofresh/stats-go/timer"
+	"sync"
```
use `goimports` sorting with groupping stdlib imports (Go -> Imports in GoLand Preferences)

vgarvardt commented on client/prometheus.go:
```diff
@@ -0,0 +1,233 @@
+		key_hash += "_" + key
+	}
+
+	if _, ok := c.increments[metric+key_hash]; !ok {
```
you should use mutex for reading/writing to the map, otherwise app will panic on concurrent access

aleksandrzhiliaev commented:
@rafaeljesus @vgarvardt @c4pone 
I changed interface a bit and added one helper to append labels to metrics, so we don't need to change or add new duplicate methods to main `Client` interface, need your opinion:

```go
operation := bucket.MetricOperation{"files", "file", "upload"}
client_prometheus.WithLabels(map[string]string{"file_type": "png", "success": "true"}).TrackMetric("file_uploader", operation)
client_prometheus.WithLabels(map[string]string{"success": "yes"}).TrackState("file_uploader_state", operation, 12345)
```

rafaeljesus commented on bucket/bucket_prometheus.go:
```diff
@@ -0,0 +1,51 @@
+
+// MetricOperationWithLabels is a struct of metric operations and labels to use for metric
+type MetricOperationWithLabels struct {
+	MetricOperation MetricOperation
```
declaring like below you have the same effect:

```go
type MetricOperationWithLabels struct {
  MetricOperation
}
```

rafaeljesus commented on bucket/bucket_prometheus.go:
```diff
@@ -0,0 +1,51 @@
+	for k, v := range operation {
+		operationSanitized[k] = SanitizeMetricName(v, uniDecode)
+	}
+	return &Prometheus{SanitizeMetricName(section, uniDecode), strings.Join(operationSanitized, "_"), success}
```
For this particular case I would declare like:

```go
return &Prometheus{
  section: SanitizeMetricName(section, uniDecode),
  operation : strings.Join(operationSanitized, "_"),
  success : success,
}
```

rafaeljesus commented on client/log.go:
```diff
@@ -26,6 +26,11 @@ func NewLog(unicode bool) *Log {
 }
 
+// WithLabels adds labels map to metrics
+func (c *Log) WithLabels(labels map[string]string) Client {
```
I would rename this to `(l *Log)`

aleksandrzhiliaev replied:
We use same var names in all that package as `c` - means client here

rafaeljesus commented on incrementer/prometheus.go:
```diff
@@ -0,0 +1,106 @@
+// Increment increments metric in prometheus
+func (i *Prometheus) Increment(metric string) {
+	if i.counter == nil {
+		i.counter = prometheus.NewCounterVec(
```
why not created it in the constructor?

aleksandrzhiliaev replied:
Because new counter is created only when we call it first time with defined metric name

rafaeljesus commented:
Returning `Client` interface in `NewClient` is not paying off anymore, we should really consider letting pkg consumers in charge of implementing interfaces along with what they need. Users would consume as follows:

statsd:
```go
import "github.com/hellofresh/stats-go/statsd"

s := statsd.New(dsn)
```

prometeus:
```go
import "github.com/hellofresh/stats-go/prometeus"

p := prometeus.New(dsn)
```

log:
```go
import "github.com/hellofresh/stats-go/log"

l := log.New(dsn)
```

This way we get rid with the anomaly of other backends having to implement prometeus handler, withLables methods

vgarvardt commented:
IMO we should leave `Client` interface - this is the main idea behind this library, to change the backend just using DSN. What we can do to achieve this is to get rid of prometheus-specific `WithLabels(labels map[string]string) Client` method and break `bucket.MetricOperation` BC - make it backend-specific (like we do with the `Timer`) or make it support labels, that will be skipped in `statsd` and `memory` backends and supported in `prometheus` and `log`. IMO it is not necessary to support all backend-specific features, this library is something ORM-like - it introduces nice interface to some common functionality and all backend-specific functions must be "hidden" in backend-implementations and used by clients when they realy want this - like using pg-specific types when using SQLAlchemy.

aleksandrzhiliaev commented:
It is easy to keep interface contract right now, for Prometheus I just added new helper `WithLabels` which will pass labels to backend, it is also fine to call prometheus methods without that helper. Lets leave interface and not break other backends compatibility.

rafaeljesus commented:
I was mentioning about getting rid of `NewClient() Client`. Yes, we should expose `Client` interface for convenience in the root folder along with a new `PrometeusClient` interface, so pkg consumers decides where to use them. The implementation should satisfy the interface, and you can get it like that:

in the root folder:

```go
type (
  Client interface {
    Inc() error
  }

  PrometeusClient interface {
    Client
    Handler() http.Handle
  }
)
```

assert that implementations satisfy the interface:

// statsd implementation
```go
type Statsd struct {}

var _ Client = (*Statsd)(nil)

// implement methods which satisfy Client interface
func (s *Statsd) Inc () error {}
```

// log implementation
```go
type Log struct {}

var _ Client = (*Log)(nil)

// implement all methods which satisfy Client interface
func (l *Log) Inc () error {}
```

// prometeus implementation
```go
type Prometeus struct {}

var _ PrometeusClient = (*Prometeus)(nil)

// implement methods which satisfy PrometeusClient interface
func (p *Prometeus) Inc () error {}
func (p *Prometeus) Handler () error {}
```

Take a look at this project as a guidance https://github.com/matryer/vice

However fell free to decide if it's worth or not at this point

vgarvardt commented on client/memory.go:
```diff
@@ -177,3 +182,8 @@ func (c *Memory) SetHTTPRequestSection(section string) Client {
+
+// Handler returns metrics endpoint for prometheus backend
+func (c *Memory) Handler() http.Handler {
+	return nil
```
we'd better return handler always - in this case handler that, for example, returns `405` status code, or return map values encoded in json

vgarvardt commented on client/memory.go:
```diff
@@ -31,6 +31,11 @@ func NewMemory(unicode bool) *Memory {
 
+// WithLabels adds labels map to metrics
+func (c *Memory) WithLabels(labels map[string]string) Client {
+	return c
```
memory may store everything, so why not store labels as well?

vgarvardt commented on client/noop.go:
```diff
@@ -87,3 +92,8 @@ func (c *Noop) SetHTTPRequestSection(section string) Client {
+
+// Handler returns metrics endpoint for prometheus backend
+func (c *Noop) Handler() http.Handler {
+	return nil
```
we'd better return handler always - in this case handler that, for example, returns `405` status code

vgarvardt commented on client/prometheus.go:
```diff
@@ -0,0 +1,225 @@
+	c.Lock()
+	defer c.Unlock()
+
+	c.labels = labels
```
This will not work. Imagine you have the following code:

```go
client, _ := stats.NewClient("prometheus://your_namespace")

// HTTP-handler that handles one request
go func() {
    client.WithLabels(map[string]string{"foo": "bar"}).TrackMetric("a1", operation)
}()

// HTTP-handler that handles another request
go func() {
    client.WithLabels(map[string]string{"baz": "qux"}).TrackMetric("a2", operation)
}()
```

in this case there is no guarantee which labels will be assigned to any of the metric - all `Track*` methods do not have any side-effects, they get some parameters, make some operations with the data and finish. This method sets global state for an object that is used in `Track*`. This is why I was suggecting breaking backward compatibility and changing `bucket.MetricOperation`.

vgarvardt commented on client/prometheus.go:
```diff
@@ -0,0 +1,225 @@
+	c.increments[metric+keyHash].IncrementWithLabels(metric, requestLabels)
+	c.increments[metricTotal+keyHash].IncrementWithLabels(metricTotal, requestLabels)
+	c.labels = map[string]string{}
+	c.Unlock()
```
defer this call right after locking, otherwise any potential problem (like panicing becuase of nil) will lock all requests

vgarvardt commented on client/statsd.go:
```diff
@@ -163,3 +168,8 @@ func (c *StatsD) SetHTTPRequestSection(section string) Client {
+
+// Handler returns metrics endpoint for prometheus backend
+func (c *StatsD) Handler() http.Handler {
+	return nil
```
we'd better return handler always - in this case handler that, for example, returns `405` status code

aleksandrzhiliaev commented:
@vgarvardt I updated interface and example in PR, please check

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,27 @@ import (
+}
+
+// NewMetricOperation  builds and returns new MetricOperation instance with defined label keys
+func NewMetricOperation(operations [3]string, labelNames []string) MetricOperation {
```
I'd change this to be something like this:

```
func NewMetricOperation(operaion ...string) *MetricOperation {
    ...
}
```

this makes new object creation less verbose and more elegant, and if want to labels - you already introduced `WithLabels(labels ...string)`, so you will get the following:

- now: `NewMetricOperation([3]string{"get", "token", "client_credentials"}, []string{})`
- after: `NewMetricOperation("get", "token", "client_credentials")`

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,27 @@ import (
+}
+
+// WithLabels adds label value to existing MetricOperation instance
+func (m *MetricOperation) WithLabels(labels ...string) MetricOperation {
```
you're receiving pointer to an object but returning an object

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,27 @@ import (
+
+// NewMetricOperation  builds and returns new MetricOperation instance with defined label keys
+func NewMetricOperation(operations [3]string, labelNames []string) MetricOperation {
+	labels := []string{"success"}
```
why do you have default label that is `success`?

vgarvardt commented on timer/timer.go:
```diff
@@ -10,4 +10,6 @@ type Timer interface {
 	// Finish writes elapsed time for metric
 	Finish(bucket string)
+	// FinishWithLabels writes elapsed time for metric
+	FinishWithLabels(bucket string, labels map[string]string)
```
I'd rather changed `Finish` method signature to accept variable list of labels, so that current calls won't be changed, something like this: `Finish(bucket string, label ...Label)`

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,41 @@ import (
 // MetricOperation is a list of metric operations to use for metric
-type MetricOperation [3]string
+type MetricOperation struct {
+	operations []string
```
it used to have fixed size and now it is free-sized - how is it handled now?

aleksandrzhiliaev replied:
true thing, we are not checking it right now. Will push changes to handle that right now

aleksandrzhiliaev commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,49 @@ import (
+	for k := range labels {
+		if _, ok := m.Labels[k]; !ok {
+			// TODO: handle error properly
+			panic("undefined labelName: " + k)
```
Any thoughts how can I raise error here in better go way?

rafaeljesus replied:
either you return the error or panic one, clients will need to recover

vgarvardt replied:
just ignore it - panicking in stats library is kinda bad idea

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,49 @@ import (
+
+// NewMetricOperation  builds and returns new MetricOperation instance with defined label keys
+func NewMetricOperation(operations ...string) MetricOperation {
+	ops := []string{"-", "-", "-"}
```
it should be `MetricEmptyPlaceholder` instead of `"-"`

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,49 @@ import (
+func NewMetricOperation(operations ...string) MetricOperation {
+	ops := []string{"-", "-", "-"}
+
+	for i := 0; i < 3; i++ {
```
3 - pls extract it to a const with explicit name to avoid creating magic constant here

vgarvardt commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,49 @@ import (
+	ops := []string{"-", "-", "-"}
+
+	for i := 0; i < 3; i++ {
+		if len(operations) <= i {
```
it would be better to do something like:
```
switch opsLen := len(operations):
case opsLen == 3: (just assign)
case opsLen > 3: (get the slice)
case opsLen < 3: (do the loop)
```

rafaeljesus commented on bucket/bucket_plain.go:
```diff
@@ -5,7 +5,49 @@ import (
+	}
+
+	for k := range labels {
+		if _, ok := m.Labels[k]; !ok {
```
We need to make it thread safe

rafaeljesus replied:
I mean we need to lock and unlock while accessing maps

rafaeljesus commented on client/prometheus.go:
```diff
@@ -0,0 +1,218 @@
+	if operation.Labels == nil {
+		operation.Labels = map[string]string{"success": strconv.FormatBool(success)}
+	} else {
+		operation.Labels["success"] = strconv.FormatBool(success)
```
we should also make lock and unlock while reading from the map

rafaeljesus commented on client/prometheus.go:
```diff
@@ -0,0 +1,218 @@
+	c.Lock()
+	defer c.Unlock()
+
+	b := bucket.NewPrometheus(section, operation, true, c.unicode)
```
Ideally it should be before acquiring the lock:

```go
b := bucket.NewPrometheus(section, operation, true, c.unicode)
metric := b.Metric()

c.Lock()
defer c.Unlock()

// code

```

rafaeljesus commented on client/prometheus.go:
```diff
@@ -0,0 +1,218 @@
+
+// TrackMetricN tracks custom metric with n diff, w/out ok/fail additional sections
+func (c *Prometheus) TrackMetricN(section string, operation bucket.MetricOperation, n int) Client {
+	c.Lock()
```
same as previous comment [r181671945](https://github.com/hellofresh/stats-go/pull/24/files#r181671945)

rafaeljesus commented on client/prometheus.go:
```diff
@@ -0,0 +1,218 @@
+
+// GetHTTPMetricCallback gets callback handler that allows metric operation alteration for HTTP Request
+func (c *Prometheus) GetHTTPMetricCallback() bucket.HTTPMetricNameAlterCallback {
+	c.Lock()
```
if I am not mistaken this lock is unnecessary

rafaeljesus commented on client/prometheus.go:
```diff
@@ -0,0 +1,218 @@
+		increments: make(map[string]incrementer.Incrementer),
+		states:     make(map[string]state.State),
+	}
+	return client, nil
```
any particular reason to return err in the constructor?

rafaeljesus commented on incrementer/prometheus.go:
```diff
@@ -0,0 +1,108 @@
+
+// IncrementAll increments all metrics for given bucket in prometheus
+func (i *Prometheus) IncrementAll(b bucket.Bucket) {
+
```
👀

rafaeljesus commented on incrementer/prometheus_test.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+func (c *CounterMock) Set(float64) {
+
```
Usually empty mock methods are structured like:

```go
func (c *CounterMock) Set(float64) {}
func (c *CounterMock) Inc() {}
func (c *CounterMock) Add(float64) {}
func (m *CounterVecMock) With(labels prometheus.Labels) prometheus.Counter { return nil }
```

rafaeljesus commented on state/prometheus_test.go:
```diff
@@ -0,0 +1,97 @@
+	addCalledValue float64
+}
+
+func (c *GaugeMock) Set(float64) {
```
same comment as [r181678806](https://github.com/hellofresh/stats-go/pull/24/files#r181678806)

rafaeljesus commented on timer/prometheus.go:
```diff
@@ -0,0 +1,52 @@
+
+	duration := time.Since(t.timerStart)
+
+	if _, ok := t.histogramVecs[bucket]; !ok {
```
should we lock/unlock here as well?

vgarvardt commented on client/prometheus.go:
```diff
@@ -51,6 +51,31 @@ func (c *Prometheus) Close() error {
 }
 
+// buildIncrement calls incrementer factory if incrementer was not created before
+func (c *Prometheus) buildIncrement(metric, metricTotal string) {
```
I'd do `getIncrement(name string)` instead to avoid accessing the map from other parts of the code, so it would be somethign like:

```golang
func (c *Prometheus) getIncrement(name string) incrementer.Incrementer {
    c.Lock()
    defer c.Unlock()

    increment, ok := c.increments[name]
    if !ok {
        increment = c.incFactory.Create()
       c.increments[metric] = increment
    }

    return increment
}
```

vgarvardt commented on client/prometheus.go:
```diff
@@ -51,6 +51,31 @@ func (c *Prometheus) Close() error {
+}
+
+// buildState calls state factory if state objects was not created before
+func (c *Prometheus) buildState(metric string) {
```
same as increment - work with the map only here, the rest should use reference to an object

rafaeljesus commented on bucket/bucket_prometheus_test.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+func TestPrometheus_Metric(t *testing.T) {
+	dataProvider := []struct {
```
👍

vgarvardt commented on timer/timer.go:
```diff
@@ -10,4 +10,5 @@ type Timer interface {
 	StartAt(time.Time) Timer
 	// Finish returns elapsed time
 	Finish() time.Duration
+	// Finish writes elapsed time for metric
```
leftover :trollface:

rafaeljesus commented:
:shipit:

julyate commented:
+1
